provider/provider.go

```diff
@@ -113,3 +113,60 @@ func pinSet(ctx context.Context, pinning pin.Pinner, fetchConfig fetcher.Factory
 
 	return set, nil
 }
+
+func NewPrioritizedProvider(priorityCids KeyChanFunc, otherCids KeyChanFunc) KeyChanFunc {
+	return func(ctx context.Context) (<-chan cid.Cid, error) {
+		outCh := make(chan cid.Cid)
+
+		go func() {
+			defer close(outCh)
+			visited := map[string]struct{}{}
+
+			handleStream := func(stream KeyChanFunc, markVisited bool) error {
+				ch, err := stream(ctx)
+				if err != nil {
+					return err
+				}
+
+				for {
+					select {
+					case <-ctx.Done():
+						return nil
+					case c, ok := <-ch:
+						if !ok {
+							return nil
+						}
+
+						// Likely faster than c.String()
+						str := string(c.Bytes())
+						if _, ok := visited[str]; ok {
+							continue
+						}
+
+						select {
+						case <-ctx.Done():
+							return nil
+						case outCh <- c:
+							if markVisited {
+								visited[str] = struct{}{}
+							}
+						}
+					}
+				}
+			}
+
+			err := handleStream(priorityCids, true)
+			if err != nil {
+				log.Warnf("error in prioritized strategy while handling priority CIDs: %w", err)
+				return
+			}
+
+			err = handleStream(otherCids, false)
```

Comment:
Is the idea that each keyFunc should do its own internal deduplication and so setting this to false saves some RAM?

Reply:
No, the point was to avoid memory issues assuming that the priorityCids stream is much smaller than the otherCids. See more context here: https://github.com/ipfs/boxo/pull/595#discussion_r1539253269

Reply:
Well, it seems that each keyFunc does indeed do its own internal deduplication, so I think I can say: yes.